### PR TITLE
feat: Update client-side telemetry

### PR DESCRIPTION
**Pull request by mildaniel**
#### Why is this change necessary?
To collect hook-specific data.

#### How does it address the issue?
Update the telemetry that is sent by sam-cli to the telemetry back end to include hook-specific data.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by moelasmar — APPROVED** [review]
It looks good to me. Thanks Daniel

File: samcli/lib/telemetry/metric.py [hunk 1]
```diff
@@ -21,7 +24,10 @@
 from samcli.lib.telemetry.event import EventTracker
 from samcli.lib.telemetry.project_metadata import get_git_remote_origin_url, get_project_name, get_initial_commit_hash
 from samcli.commands._utils.experimental import get_all_experimental_statues
+from samcli.lib.hook.hook_config import HookPackageConfig
+from samcli.lib.hook.hook_wrapper import INTERNAL_PACKAGES_ROOT
 from .telemetry import Telemetry
+from ..hook.exceptions import InvalidHookPackageConfigException
```
Comment (hawflau) [inline samcli/lib/telemetry/metric.py:30]: nit: should we change all relative imports to absolute imports?

File: samcli/lib/telemetry/metric.py [hunk 2]
```diff
@@ -185,6 +202,22 @@ def wrapped(*args, **kwargs):
...
 
 
+def _get_project_details(hook_package_id: str, template_dict: Dict) -> ProjectDetails:
+    if not hook_package_id:
+        project_type = ProjectTypes.CDK.value if is_cdk_project(template_dict) else ProjectTypes.CFN.value
+        return ProjectDetails(project_type=project_type, hook_package_id=None, hook_package_version=None)
+    hook_location = Path(INTERNAL_PACKAGES_ROOT, hook_package_id)
+    try:
+        hook_package_config = HookPackageConfig(package_dir=hook_location)
+    except InvalidHookPackageConfigException:
```
Comment (hawflau) [inline samcli/lib/telemetry/metric.py:212]: The way we now invoke the hook (via a click option class), if it fails, seems like it won't  trigger the `track_command` method. 

So this seems like an unlikely path that will get triggered.

But we should think about how to capture that kind of failures and make sure we report them via Telemetry.

Comment (mildaniel) [reply samcli/lib/telemetry/metric.py:212]: We discussed this point offline. We generally don't send telemetry for bad click options which is a miss on our end. We will create a ticket internally to track this and follow up.

File: samcli/lib/hook/hook_config.py [hunk 3]
```diff
@@ -87,3 +87,7 @@ def description(self) -> Optional[str]:
     @property
     def functionalities(self) -> Dict[str, HookFunctionality]:
         return cast(Dict[str, HookFunctionality], self._config["functionalities"])
+
+    @property
+    def iac_framework(self) -> str:
+        return cast(str, self._config.get("iac_framework") if self._config.get("iac_framework") else "")
```
Comment (hawflau) [inline samcli/lib/hook/hook_config.py:93]: Would this be the same?
```suggestion
        return cast(str, self._config.get("iac_framework", ""))
```

**Review by hawflau — APPROVED** [review]
approved with a minor comment

File: samcli/lib/telemetry/metric.py [hunk 4]
```diff
@@ -1,29 +1,35 @@
...
-from typing import Optional, Tuple
+from typing import Optional, Tuple, Dict
 
 import click
 
 from samcli import __version__ as samcli_version
 from samcli.cli.context import Context
 from samcli.cli.global_config import GlobalConfig
-from samcli.lib.warnings.sam_cli_warning import TemplateWarningsChecker
+from samcli.commands._utils.experimental import get_all_experimental_statues
```
Comment (sriram-mv) [inline samcli/lib/telemetry/metric.py:21]: Nothing to do with this particular PR.

but note regardless: this file gets imported a LOT across all commands, so we need to be mindful of the imports we put in here and where possible rely on lazy-imports.

**Review by sriram-mv — APPROVED** [review]

**Pull request merged**